console-subscriber/src/lib.rs

```diff
@@ -957,9 +957,110 @@ impl Server {
+    /// other endpoints from the same gRPC server.
+    ///
+    /// ```
+    /// use console_subscriber::{ConsoleLayer, ServerParts};
+    ///
+    /// # let runtime = tokio::runtime::Builder::new_current_thread()
+    /// #     .enable_all()
+    /// #     .build()
+    /// #     .unwrap();
+    /// # runtime.block_on(async {
+    /// let (console_layer, server) = ConsoleLayer::builder().build();
+    /// let ServerParts {
+    ///     instrument_server,
+    ///     aggregator_handle,
+    ///     ..
+    /// } = server.into_parts();
+    ///
+    /// let router = tonic::transport::Server::builder()
+    ///     //.add_service(some_other_service)
+    ///     .add_service(instrument_server);
+    /// let serve = router.serve(std::net::SocketAddr::new(
+    ///     std::net::IpAddr::V4(std::net::Ipv4Addr::new(127, 0, 0, 1)),
+    ///     6669,
+    /// ));
+    ///
+    /// // Finally, spawn the server.
+    /// tokio::spawn(serve);
+    /// # // Avoid a warning that `console_layer` and `aggregator_handle` are unused.
+    /// # drop(console_layer);
+    /// # drop(aggregator_handle);
+    /// # });
+    /// ```
+    ///
+    /// [`Router::serve`]: fn@tonic::transport::server::Router::serve
+    pub fn into_parts(mut self) -> ServerParts {
+        let aggregate = self
+            .aggregator
+            .take()
+            .expect("cannot start server multiple times");
+        let aggregate = spawn_named(aggregate.run(), "console::aggregate");
+
+        let service = proto::instrument::instrument_server::InstrumentServer::new(self);
+
+        ServerParts {
+            instrument_server: service,
+            aggregator_handle: AggregatorHandle {
+                join_handle: aggregate,
+            },
+        }
+    }
+}
+
+/// Server Parts
+///
+/// This struct contains the parts returned by [`Server::into_parts`]. It may contain
+/// further parts in the future, an as such is marked as `non_exhaustive`.
+///
+/// The `InstrumentServer<Server>` can be used to construct a router which
+/// can be added to a [`tonic`] gRPC server.
+///
+/// The [`AggregatorHandle`] must be kept until after the server has been
+/// shut down.
+///
+/// See the [`Server::into_parts`] documentation for usage.
+#[non_exhaustive]
+pub struct ServerParts {
+    /// The instrument server.
+    ///
+    /// See the documentation for [`InstrumentServer`] for details.
+    pub instrument_server: InstrumentServer<Server>,
+
+    /// The aggregate handle.
+    ///
+    /// See the documentation for [`AggregatorHandle`] for details.
+    pub aggregator_handle: AggregatorHandle,
+}
+
+/// Aggregator handle.
+///
+/// This object is returned from [`Server::into_parts`] and must be
+/// kept as long as the `InstrumentServer<Server>` - which is also
+/// returned - is in use.
+pub struct AggregatorHandle {
+    join_handle: JoinHandle<()>,
+}
+
+impl Drop for AggregatorHandle {
+    fn drop(&mut self) {
+        self.join_handle.abort();
+    }
```

Comment:
I'm not convinced we should always do this. Instead, perhaps, the `ServerParts` should just give the user access to the `JoinHandle` so they can decide whether they want to abort the task? The behavior  of `serve_with` was specifically intended to ensure that the aggregator is shut down if the server task ends.

Reply:
@hawkw OK, that makes sense. I was hoping to avoid abstract away the fact that this is a join handle underneath. Do you think there's any value in that?

I see 2 options here:
* Keep the `AggregatorHandle` without the `impl Drop` and add an `abort()` method to explicitly shut down the aggregator.
* Expose the join handle as is.

Do you have a specific preference?

Reply:
@hawkw i went with the first option. Let me know what you think.

Reply:
> OK, that makes sense. I was hoping to avoid abstract away the fact that this is a join handle underneath

Hmm...now that I think about it a bit more, wouldn't the ideal lower-level interface just be to expose the future itself, instead of spawning it and handing the user a `JoinHandle`? Then, the user could spawn the aggregator task wherever they want to, which could be useful if the user is using a non-Tokio runtime, or has multiple Tokio runtime, and wants to control where the aggregator task is spawned. Also, we can avoid having to think about the `JoinHandle` entirely if it's up to the user to spawn the aggregator. 

What do you think?